File: java/Sorts/src/InsertionSort.java
/**
 * Iterate from 0 to n-1 inserting elements in their proper position by
 * comparing ith element to ith+1 element using a greater than comparison.
 * Worst case O(N^2)
 * @author bradley
 *
 */
class InsertionSort {
	private static <T extends Comparable<T>> void insert(T[] source, int rightIndex) 
	{
		T value = source[rightIndex + 1];
		
		while(rightIndex >= 0 && source[rightIndex].compareTo(value) > 0) {
			source[rightIndex + 1] = source[rightIndex--];
		}
		
		source[rightIndex + 1] = value;
	}
	
	static <T extends Comparable<T>> void sort(T[] source) 
	{
		int nMinusOne = source.length - 1;
		
		for(int i=0; i < nMinusOne; i++) {
			insert(source, i);
		}
	}
}

File: csharp/Sorts/InsertionSort.cs
using System;

namespace Sorts
{
	/// <summary>
	/// Iterate from 0 to n-1 inserting elements in their proper position by
	/// comparing ith element to ith+1 element using a greater than comparison.
	/// Worst case O(N^2)
	/// </summary>
	public class InsertionSort
	{
		private static void Insert<T>(T[] array, int rightIndex) where T : IComparable
		{
			T value = array [rightIndex + 1];

			while (rightIndex >= 0 && array[rightIndex].CompareTo(value) > 0) {
				array [rightIndex + 1] = array [rightIndex--];
			}

			array[rightIndex + 1] = value;
		}

		public static void Sort<T>(T[] array) where T : IComparable
		{
			int nMinusOne = array.Length - 1;
			for (int i = 0; i < nMinusOne; i++) {
				Insert (array, i);
			}
		}
	}
}

